### ISSUE TITLE
No Debug Symbols when using per_file_copt option

### ISSUE THREAD

**Issue by makn87** [tensorflow/tensorflow#40130]
<em>Please make sure that this is a build/installation issue. As per our [GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md), we only address code/doc bugs, performance issues, feature requests and build/installation issues on GitHub. tag:build_template</em>

**System information**
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): RHEL 8 (Linux 4.18.0-147)
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device:
- TensorFlow installed from (source or binary): source
- TensorFlow version: 2.2.0
- Python version: 3.7
- Installed using virtualenv? pip? conda?:virtualenv
- Bazel version (if compiling from source): 2.0.0
- GCC/Compiler version (if compiling from source): 8.3.1
- CUDA/cuDNN version: Not used
- GPU model and memory: Not used

I've been building 2.2.0 from source (targets:build_pip_package and libtensorflow_cc.so) while  selectively including debug flags, using the per_file_copt option (per suggestion in issue #27495). 

 My build succeeds, but the debug symbols aren't there for either libtensorflow_cc.so.2.2.0 or the libtensorflow_framework.so.2.2.0.

 Below are some of the options I've tried. If I examine the .so files, with 'readelf -S libtensorflow_cc.so.2.2.0' they indicate that the debug symbols exist, but not so for libtensorflow_framework.so.2.2.0. TotalView debugger and gdb don't appear to find any of the debug symbols

**Provide the exact sequence of commands / steps that you executed before running into the problem**

I always build the build_pip_package target first. After it's complete I provide the exact same command (below), changing the build target to libtensorflow_cc.so

Variation 1 (This doesn't uses the per_file_copt option, but demonstrates that the --strip=never compile option gives an error when put after the '-c' flags):
 bazel build -c dbg --strip=never //tensorflow/tools/pip_package:build_pip_package

Variation 2: (moving the --strip=never flag in front of the -c options)
 bazel build --strip=never -c opt --config=opt --per_file_copt=//tensorflow/c/.*\.cc,//tensorflow/cc/.*/.*\.cc,//tensorflow/core/.*/.*\.cc@-g //tensorflow/tools/pip_package:build_pip_package

Variation 3 (try to prevent stripping by passing -O0 to the per_file_copt)
bazel build -c opt --config=opt --per_file_copt=//tensorflow/c/.*\.cc,//tensorflow/cc/.*/.*\.cc,//tensorflow/core/.*/.*\.cc@-g,O0  //tensorflow/tools/pip_package:build_pip_package

Please advise as to what I'm doing wrong and how to make sure that the debug symbols (which I believe are built), don't get stripped during linking.

Thanks

**Any other info / logs**
Include any logs or source code that would be helpful to diagnose the problem. If including tracebacks, please include the full traceback. Large logs and files should be attached.

**Comment by makn87** [comment]
Need to verify something in second build target

**Comment by makn87** [comment]
Had a typo in target name, should have commented inside of closed issue

**Comment by gunan** [comment]
I myself never sucessfully build TF with debug symbols.
@hlopko could you comment on the options used above to build with bazel with debug symbols?

**Comment by makn87** [comment]
This is a very difficult problem for me. I'm trying to advocate that my company migrate their software to TF 2.X, from 1.X. We use the C/C++ API for inference and Python API for model training. 

We've observed threading contention and other issues when calling TF C++ API from within our product. So, I'd really like to be able to do some debugging. If I compile the entire program with debug symbols, the .so file is > 4G and I can't even link to it. So unfortunately that's not an option. 

I need debug symbols for a handful of files. Thanks

**Comment by byronyi** [comment]
You could take a look at https://github.com/tensorflow/tensorflow/issues/27495#issuecomment-481101932

**Comment by makn87** [comment]
Thanks. I did look at that, and followed it very closely. It's cross-referenced in my original post.

 I had the same issues as that user. I also got desperate and decided to compile with the  -g flag set for all files.  After compiling, the linker was trying to link for hours so I gave up. In addition to the fact that the .so file was > 15G, compiling a full debug build seems unhelpful for me. I can't link it.

Since my original post I've tried a few additional experiments. For example, I found out that the option _--strip=never_ needs to passed to the bazel build command before an options that have a '-c' or '-copt' flag. But, unfortunately this didn't help.

In issue #27495, the individual who submitted the ticket made the following statement:
> with this new command, gdb can read symbols from libtensorflow.so, but still unable to read from libtensorflow_framework.so

I'm not sure how this was determined, but if I run 
`readelf -S bazel-bin/tensorflow/libtensorflow.so.2.2.0`, 
I see lines matching the pattern _.debug\_\***_

If I run 
`readelf -S bazel-bin/libtensorflow_framework.so.2.2.0`
I don't see any lines that match the pattern _.debug\_\***_

This seems to confirm the findings in #27495.

So maybe the fact that I build the _libtensorflow.so_ target after the _build_pip_package and libtensorflow_cc.so_ target is causing my debug symbols to be stripped when _libtensorflow.so_ is linked?

What is built by the target _libtensorflow.so_ that isn't built using the target _build_pip_package and libtensorflow_cc.so_ ? 

It seems that when I build _libtensorflow.so_ after building _build_pip_package and libtensorflow_cc.so_, all that is done is some linking.

I have used bazel query "deps(***)" to find the dependencies of these libraries. Is there anything in more human readable format that describes what the different targets build, or do I need to read through all of the BUILD files?

Thanks

**Comment by hlopko** [comment]
Hi all, I apologize for not investigating deeper (I'm on vacation and in the process of moving to a different city), so I'll only throw some breadcrumbs that might help.

To tell Bazel you want to build in the debug mode use [--compilation_mode dbg](https://docs.bazel.build/versions/master/command-line-reference.html#flag--compilation_mode) or `-c dbg`. 

If you only want to include debug symbols for some files, use [--per_file_copt](https://docs.bazel.build/versions/master/command-line-reference.html#flag--per_file_copt).

And for stripping, I'm not sure which C++ toolchain are you using, but the one built into Bazel will strip debug info [during linking](https://cs.opensource.google/bazel/bazel/+/master:tools/cpp/unix_cc_toolchain_config.bzl;l=711) depending on the value of the [--strip](https://docs.bazel.build/versions/master/command-line-reference.html#flag--strip). Put `never` there to be sure.

**Comment by makn87** [comment]
hlopko,
Thanks for replying, especially since you are on vacation. 

I think from you post, I can see what I may have done incorrectly.

Originally I put the '--strip=never' after -c dbg, but got an error because of the ordering. 

I thought that by doing 
`bazel build --strip=never -c opt --config=opt --per_file_copt=//tensorflow/c/..cc,//tensorflow/cc/./..cc,//tensorflow/core/./.*.cc@-g //tensorflow/tools/pip_package:build_pip_package`

That the '-g' flag would be applied to all files matching my regex

But seems I should have done
`bazel build --strip=never -c dbg --per_file_copt=//tensorflow/c/..cc,//tensorflow/cc/./..cc,//tensorflow/core/./.*.cc //tensorflow/tools/pip_package:build_pip_package` 

I'll give this a try. I'm using bazel 2.0.0, that I built from source. My gcc is 8.3.1.

**Comment by makn87** [comment]
OK

This doesn't work. 
`bazel build -c dbg --strip=never --per_file_copt=//tensorflow/c/.*\.cc,//tensorflow/cc/saved_model/.*\.cc,//tensorflow/core/framework/.*\.cc,//tensorflow/cc/framework/.*\.cc //tensorflow/tools/pip_package:build_pip_package`

Using the command above for either the target `build_pip_package` or `libtensorflow_cc.so` , results in the following error:

` external/aws-checksums/source/intel/crc32c_sse42_asm.c: In function 's_crc32c_sse42_clmul_256':
external/aws-checksums/source/intel/crc32c_sse42_asm.c:61:5: error: 'asm' operand has impossible constraints
     asm volatile("enter_256_%=:"
     ^~~
Target //tensorflow/tools/pip_package:build_pip_package failed to build `

This is apparently issue #37498 

I've combed over the internet for months trying to find a way to build with debug symbols for select files. I also posted a question on stack overflow months ago.  I haven't come across a solid answer yet. 

It seems that @gunan's response is fairly common. Many people, Google employees included,  haven't built TF with debug symbols. 

I've only been developing software in industry for 10 years, but I write code every day that has mistakes. I can't imagine trying to debug my code without a debugger. In fact, the fancier the debugger, the more I like it. For a while my debugger of choice has been TotalView.  

I know you guys are all top developers, but how do you debug code if you don't build it with debug symbols? This isn't meant to be condescending, I'm really just curious :) . 

Is the only way forward to use the work around suggested in that ticket?

**Comment by byronyi** [comment]
I would suggest you to use -c opt with the per file copt option.

**Comment by makn87** [comment]
Can you please elaborate on how this is different from what I tried and listed as 'Variation 2' and 'Variation 3' in the original ticket submission?

**Comment by byronyi** [comment]
```//tensorflow/c/..cc,//tensorflow/cc/./..cc,//tensorflow/core/./.*.cc```
 is not a valid path filter.

See https://docs.bazel.build/versions/master/user-manual.html#flag--per_file_copt for an example.

PS: you could try ```//tensorflow/c/.*\.cc,//tensorflow/cc/.*\.cc,//tensorflow/core/.*\.cc@-g,-O0```

**Comment by makn87** [comment]
I'm sorry @byronyi,

Somehow the path filter I posted in my original post got corrupted when I pasted it.

I did have the `.` properly escaped in my bazel build command. This filter path I used is below, I got this from the  bazel user manual. I should have the proper level of recursion. 
`--per_file_copt=//tensorflow/c/.*\.cc,//tensorflow/cc/.*/.*\.cc,//tensorflow/core/.*/.*\.cc`

I'll try your path filter, but I'd think that at least `//tensorflow/c/.*\.cc` should have produced debug symbols for the .cc files in the C API. Thanks

**Comment by makn87** [comment]
@byronyi ,

I used your filter path. The debugger definitely loaded some debug symbols, but there was just nothing but addresses when I stepped through my source.

I do have a habit of changing my home directory temporarily when I do bazel build, to keep from filling up my home directory with its .cache (which is a far smaller size then the drive where I put the installs). I know there is an option to set the cache directory elsewhere, but I've never had luck with this.
